session/kx.go

```diff
@@ -0,0 +1,439 @@
+// Copyright (c) 2016,2017 Company 0, LLC.
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package session
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/sha512"
+	"errors"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/companyzero/sntrup4591761"
+	"github.com/davecgh/go-xdr/xdr2"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+var (
+	ErrDecrypt   = errors.New("decrypt failure")
+	ErrOverflow  = errors.New("message too large")
+	ErrInvalidKx = errors.New("invalid kx")
+	ErrMarshal   = errors.New("could not marshal")
+	ErrUnmarshal = errors.New("could not unmarshal")
+)
+
+// KX allows two peers to derive a pair of shared keys. One peer must trigger
+// Initiate (the client) while the other (the server) should call Init once
+// followed by Respond for each connection.
+type KX struct {
+	Conn           net.Conn
+	MaxMessageSize uint
+	OurPrivateKey  *[sntrup4591761.PrivateKeySize]byte
+	OurPublicKey   *[sntrup4591761.PublicKeySize]byte
+	TheirPublicKey *[sntrup4591761.PublicKeySize]byte
+	writeKey       *[32]byte
+	readKey        *[32]byte
+	writeSeq       [24]byte
+	readSeq        [24]byte
+}
+
+// A pair of ephemeral keys is kept to ensure key erasure (forward secrecy)
+// should long-term keys be compromised.
+var (
+	ephemeralPublic  [sntrup4591761.PublicKeySize]byte
+	ephemeralPrivate [sntrup4591761.PrivateKeySize]byte
+	ephemeralMutex   sync.Mutex
+)
+
+// regenerateEphemeral rotates the server/client's ephemeral key. It is invoked
+// concurrently to the operation of the server/client, therefore we need to
+// acquire a mutex to ensure noninterference.
+func regenerateEphemeral() error {
+	pk, sk, err := sntrup4591761.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	ephemeralMutex.Lock()
+	copy(ephemeralPublic[:], pk[:])
+	copy(ephemeralPrivate[:], sk[:])
+	ephemeralMutex.Unlock()
```

Comment:
please use defer on locks and we really want a context that carries said lock.

Reply:
Hi Marco,

I am not sure if I understood you.

This lock exists in the scope of the session module. I can have it
encapsulated inside another variable in the same scope, but what would
that change?

Regarding 'defer', this lock is released in two places. The first is
just before a 'return', in which case 'defer' would change nothing. The
second is in Respond, where we _want_ to release the lock early in the
function to increase parallelism.

Releasing the lock explicitly also helps to make it clear what is
protected by the lock.